waspc/waspls/src/Wasp/LSP/CodeActions.hs

```diff
@@ -0,0 +1,160 @@
+import qualified Wasp.LSP.TypeInference as Inference
+import Wasp.LSP.Util (absFileInProjectRootDir)
+import Wasp.Util.IO (doesFileExist)
+
+-- | Runs all 'codeActionProviders' and concatenates their results.
+getCodeActionsInRange :: LSP.Range -> HandlerM [LSP.CodeAction]
+getCodeActionsInRange range = do
+  src <- asks (^. State.currentWaspSource)
+  maybeCst <- asks (^. State.cst)
+  -- VSCode sends codeAction requests with ranges where the start
+  -- position is equal to the end position. This range contains 0 characters,
+  -- so we add a character to it.
+  --
+  -- The LSP specification specifies that the end of a range is exclusive:
+  -- https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#range
+  let sourceSpan =
+        let SourceSpan s e = lspRangeToSpan src range
+         in if s == e then SourceSpan s (e + 1) else SourceSpan s e
+  case maybeCst of
+    Nothing -> pure []
+    Just syntax -> do
+      concat <$> mapM (\f -> f src syntax sourceSpan) codeActionProviders
+
+codeActionProviders :: [String -> [SyntaxNode] -> SourceSpan -> HandlerM [LSP.CodeAction]]
+codeActionProviders =
+  [ tsScaffoldActionProvider
+  ]
+
+-- | Provide 'LSP.CodeAction's for missing external imports. The code actions
+-- run the "Wasp.LSP.Commands.ScaffoldTsSymbol" command to define the missing
+-- function in a JS/TS file.
+tsScaffoldActionProvider :: String -> [SyntaxNode] -> SourceSpan -> HandlerM [LSP.CodeAction]
+tsScaffoldActionProvider src syntax sourceSpan = do
+  let extImports = map (extImportAtLocation src) $ collectExtImportNodesInSpan (T.fromSyntaxForest syntax)
+  concat <$> mapM (findCodeActionsForExtImport src) extImports
+  where
+    -- Post-condition: all 'Traversal's returned have @kindAt t == ExtImport@.
+    collectExtImportNodesInSpan :: Traversal -> [Traversal]
+    collectExtImportNodesInSpan t =
+      -- Only consider the given traversal if it is within the span.
+      if spansOverlap (T.spanAt t) sourceSpan
+        then case T.kindAt t of
+          S.ExtImport -> [t]
+          _ -> concatMap collectExtImportNodesInSpan $ T.children t
+        else []
+
+-- | Finds code actions to create a JS/TS function for an external import. Returns
+-- one code action for each JS/TS/JSX/TSX file a function can be created in.
+--
+-- This function also checks to make sure each code action, which runs the
+-- @wasp.scaffold.ts-symbol@ command, has a scaffolding template that can be
+-- used.
+findCodeActionsForExtImport :: String -> ExtImportNode -> HandlerM [LSP.CodeAction]
+findCodeActionsForExtImport src extImport = do
+  -- 1. Lookup up the extimport in the exports cache to see if it is valid.
+  -- 2. If it is, then no code actions need to be returned.
+  -- 3. Otherwise, path through the expression tree to the ext import.
+  -- 4. Use the extension type stored in the cache key to get a list of possible
+  --    files the user might want to add the code to.
```

Comment:
Ok yes, this helps for sure!

I suggest, if nothing else, also putting this in the comment in the code.

Of course, if you can express it via code instead of comments that would be even better, but it sounds like this will end up in the comments, as an example / specific situation.